cmd/src/main.go

```diff
@@ -157,6 +169,60 @@ func readConfig() (*config, error) {
 	if cfg.Endpoint == "" {
 		cfg.Endpoint = "https://sourcegraph.com"
 	}
+	if envProxyEndpoint != "" {
+		cfg.ProxyEndpoint = envProxyEndpoint
+	}
+
+	if cfg.ProxyEndpoint != "" {
+
+		parseEndpoint := func(endpoint string) (scheme string, address string) {
+			parts := strings.SplitN(endpoint, "://", 2)
+			if len(parts) == 2 {
+				return parts[0], parts[1]
+			}
+			return "", endpoint
+		}
+
+		urlSchemes := []string{"http", "https", "socks", "socks5", "socks5h"}
+
+		isURLScheme := func(scheme string) bool {
+			for _, s := range urlSchemes {
+				if scheme == s {
+					return true
+				}
+			}
+			return false
+		}
+
+		scheme, address := parseEndpoint(cfg.ProxyEndpoint)
+
+		if isURLScheme(scheme) {
+			endpoint := cfg.ProxyEndpoint
+			// assume socks means socks5, because that's all we support
+			if scheme == "socks" {
+				endpoint = "socks5://" + address
+			}
+			cfg.ProxyEndpointURL, err = url.Parse(endpoint)
+			if err != nil {
+				return nil, err
+			}
+		} else if scheme == "" || scheme == "unix" {
+			path, err := expandHomeDir(address)
+			if err != nil {
+				return nil, err
+			}
+			isValidUDS, err := isValidUnixSocket(path)
+			if err != nil {
+				return nil, err
+			}
+			if !isValidUDS {
+				return nil, errors.Newf("invalid proxy socket: %s", path)
+			}
```

Comment:
The intent was to provide a better error message.

Dial failures result in errors like, "dial unix /foo/bar/path: connect: no such file or directory", or "connect: socket operation on non-socket" and I wanted to provide something better up front.

I didn't completely deliver on that because the error messages during config validation aren't great either. I can make the config validation error messages better, but also, I'm not stuck on doing this validation up front.

Reply:
Ah, got it. Makes sense! No real preference here, feel free to ignore